core/client/grpcc/grpc_client.go

```diff
@@ -225,70 +223,162 @@ func (c *Client) ListHistory(ctx context.Context, userID, appID string, start, e
+		return nil, err
 	}
 
-	err = c.Retry(ctx, m, signers, opts...)
-	// Retry submitting until an inclusion proof is returned.
-	for i := 0; err == ErrRetry && i < c.RetryCount; i++ {
-		time.Sleep(c.RetryDelay)
-		err = c.Retry(ctx, m, signers, opts...)
+	// 3. Wait for update
+	m, err = c.WaitForUserUpdate(ctx, m)
+	for i := 0; i < c.RetryCount; i++ {
+		switch err {
+		case ErrWait:
+			// Try again.
+		case ErrRetry:
+			if err := c.QueueMutation(ctx, m, signers); err != nil {
+				return nil, err
+			}
+		default:
+			return m, err
+		}
+		m, err = c.WaitForUserUpdate(ctx, m)
 	}
 	return m, err
 }
 
-// Retry takes take a mutation, signs, and sends it again, and updates the back pointer with the current leaf value.
-func (c *Client) Retry(ctx context.Context, m *entry.Mutation, signers []signatures.Signer, opts ...grpc.CallOption) error {
-	req, err := m.SerializeAndSign(signers, c.trusted.TreeSize)
+// QueueMutation signs m and sends it to the server.
+func (c *Client) QueueMutation(ctx context.Context, m *entry.Mutation, signers []signatures.Signer) error {
+	req, err := m.SerializeAndSign(signers, c.trusted.GetTreeSize())
 	if err != nil {
 		return fmt.Errorf("SerializeAndSign(): %v", err)
 	}
 
 	Vlog.Printf("Sending Update request...")
-	updateResp, err := c.cli.UpdateEntry(ctx, req, opts...)
+	// 2. Queue Mutation
+	_, err = c.cli.UpdateEntry(ctx, req)
+	return err
+}
+
+// newMutation fetches the current index and value for a user and prepares a mutation.
+func (c *Client) newMutation(ctx context.Context, u *tpb.User) (*entry.Mutation, error) {
+	e, err := c.VerifiedGetEntry(ctx, u.AppId, u.UserId)
 	if err != nil {
-		return fmt.Errorf("cli.UpdateEntry(): %v", err)
+		return nil, err
 	}
+	oldLeaf := e.GetLeafProof().GetLeaf().GetLeafValue()
 	Vlog.Printf("Got current entry...")
 
-	// Validate response.
-	if err := c.kt.VerifyGetEntryResponse(ctx, c.domainID, req.AppId, req.UserId, &c.trusted, updateResp.GetProof()); err != nil {
-		return fmt.Errorf("VerifyGetEntryResponse(): %v", err)
+	index, err := c.kt.Index(e.GetVrfProof(), u.DomainId, u.AppId, u.UserId)
+	if err != nil {
+		return nil, err
 	}
 
-	cntLeaf := updateResp.GetProof().GetLeafProof().GetLeaf().GetLeafValue()
-	equal, err := m.Check(cntLeaf)
+	mutation := entry.NewMutation(index, u.DomainId, u.AppId, u.UserId)
+
+	if err := mutation.SetPrevious(oldLeaf, true); err != nil {
+		return nil, err
+	}
+
+	if err := mutation.SetCommitment(u.PublicKeyData); err != nil {
+		return nil, err
+	}
+
+	if len(u.AuthorizedKeys) != 0 {
+		if err := mutation.ReplaceAuthorizedKeys(u.AuthorizedKeys); err != nil {
+			return nil, err
+		}
+	}
+
+	return mutation, nil
+}
+
+// WaitForUserUpdate waits for the STH to be updated, indicating the next epoch has been created,
+// it then queries the current value for the user and checks it against the requested mutation.
+// If the current value has not changed, WaitForUpdate returns ErrWaitSomeMore.
```

Comment:
ErrNotReady? or similar.

Reply:
`ErrWait` just means that the update hasn't been processed yet.